lib/src/sliver_infinite_list.dart

```diff
@@ -0,0 +1,299 @@
+  }
+
+  void detachFromPosition() {
+    scrollPosition?.removeListener(attemptFetch);
+  }
+
+  void attemptFetch() {
+    if (isAtEnd &&
+        !widget.hasReachedMax &&
+        !widget.isLoading &&
+        !widget.hasError) {
+      debounce(widget.onFetchData);
+    }
+  }
+
+  bool get isAtEnd {
+    if (widget.itemCount == 0) {
+      return true;
+    }
+
+    final scrollPosition = this.scrollPosition;
+    if (scrollPosition == null) {
+      return false;
+    }
+
+    // This considers the end of the scrollable content as the
+    // position to trigger a data fetch. It may cause unintuitive behaviors
+    // when there is any sliver after this one.
+    final maxScroll = scrollPosition.maxScrollExtent;
+
+    final currentScroll = scrollPosition.pixels - widget.precedingScrollExtent;
+    return currentScroll >= (maxScroll - widget.scrollExtentThreshold);
+  }
+
+  WidgetBuilder get loadingBuilder =>
+      widget.loadingBuilder ?? defaultInfiniteListLoadingBuilder;
+
+  WidgetBuilder get errorBuilder =>
+      widget.errorBuilder ?? defaultInfiniteListErrorBuilder;
+
+  @override
+  Widget build(BuildContext context) {
+    final hasItems = widget.itemCount != 0;
+
+    final showEmpty = !widget.isLoading &&
+        widget.itemCount == 0 &&
+        widget.emptyBuilder != null;
+    final showBottomWidget = showEmpty || widget.isLoading || widget.hasError;
+    final showSeparator = widget.separatorBuilder != null;
+    final separatorCount = !showSeparator ? 0 : widget.itemCount - 1;
+
+    final effectiveItemCount =
+        (!hasItems ? 0 : widget.itemCount + separatorCount) +
+            (showBottomWidget ? 1 : 0);
+    final lastItemIndex = effectiveItemCount - 1;
+
+    return SliverList(
+      delegate: SliverChildBuilderDelegate(
+        childCount: effectiveItemCount,
+        (context, index) {
+          if (index == lastItemIndex && showBottomWidget) {
+            if (widget.hasError) {
+              return errorBuilder(context);
+            } else if (widget.isLoading) {
+              return loadingBuilder(context);
+            } else {
+              return widget.emptyBuilder!(context);
+            }
+          } else {
+            if (showSeparator && index.isOdd) {
+              return widget.separatorBuilder!(context);
+            } else {
+              final itemIndex = !showSeparator ? index : (index / 2).floor();
+              return widget.itemBuilder(context, itemIndex);
+            }
+          }
+        },
+      ),
+    );
+  }
+}
+
+/// {@template callback_debouncer}
+/// A model used for debouncing callbacks.
+///
+/// Is only used internally and should not be used explicitly.
+/// {@endtemplate}
+@visibleForTesting
+class CallbackDebouncer {
+  /// {@macro callback_debouncer}
+  CallbackDebouncer(this._delay);
+
+  final Duration _delay;
+  Timer? _timer;
+
+  /// Calls the given [callback] after the given duration has passed.
+  @visibleForTesting
```

Comment:
is this necessary if the class is already visibleForTesting?

Reply:
lol, Now I realised that. It is like that on master. Since the diff is already not considering the previous implementation, I shall fix that here.